File: utils/ProtobufTransformer.py
import Messages_pb2
from Messages_pb2 import MessageType as m_type
from google.protobuf.json_format import MessageToDict



def encode_msg(data_dict:dict, msg_type:int):
    msg_wrapper = Messages_pb2.Wrapper()
    msg_wrapper.type = msg_type

    if msg_type == m_type.SYSTEM:
        msg = Messages_pb2.System()
        msg.whoami, msg.status  = data_dict["whoami"], data_dict["status"]
        msg_wrapper.sys_msg.CopyFrom(msg)

    elif msg_type == m_type.INFORMATION:
        msg = Messages_pb2.Information()
        msg.info = data_dict["info"]
        msg_wrapper.info_msg.CopyFrom(msg) 
    
    elif msg_type == m_type.MOTORS:
        msg = Messages_pb2.Motors()
        msg.alpha, msg.beta = data_dict["alpha"], data_dict["beta"]
        msg_wrapper.motors_msg.CopyFrom(msg)

    elif msg_type == m_type.TRANSMIT:
        msg = Messages_pb2.Transmit()
        msg.frequency, msg.power = data_dict["frequency"], data_dict["power"]
        msg.wavetype, msg.onoff = data_dict["wavetype"], data_dict["onoff"]
        msg_wrapper.tx_msg.CopyFrom(msg)

    elif msg_type == m_type.SET_FREQUENCY:
        msg = Messages_pb2.SetFrequency()
        msg.frequency = data_dict["frequency"]
        msg_wrapper.freq_msg.CopyFrom(msg)   

    elif msg_type == m_type.SAMPLING:
        msg = Messages_pb2.Sampling()
        msg.onoff = data_dict["onoff"]
        msg_wrapper.sample_msg.CopyFrom(msg)
    
    elif msg_type == m_type.GET_RSSI:
        msg = Messages_pb2.GetRssi()
        msg.rssi = data_dict["rssi"]
        msg.samples = data_dict["samples"]
        msg.std = data_dict["std"]
        msg_wrapper.rssi_msg.CopyFrom(msg)

    return msg_wrapper.SerializeToString()
        
def decode_msg(serialized_data):
    msg_wrapper = Messages_pb2.Wrapper()
    msg_wrapper.ParseFromString(serialized_data)
    msg_type = int(repr(msg_wrapper.type))

    if msg_type == m_type.SYSTEM:
        msg_contents = MessageToDict(msg_wrapper.sys_msg)

    elif msg_type == m_type.INFORMATION:
        msg_contents = MessageToDict(msg_wrapper.info_msg)

    elif msg_type == m_type.MOTORS:
        msg_contents = MessageToDict(msg_wrapper.motors_msg)

    elif msg_type == m_type.TRANSMIT:
        msg_contents = MessageToDict(msg_wrapper.tx_msg)

    elif msg_type == m_type.SET_FREQUENCY:
        msg_contents = MessageToDict(msg_wrapper.freq_msg)

    elif msg_type == m_type.SAMPLING:
        msg_contents = MessageToDict(msg_wrapper.tx_msg)

    elif msg_type == m_type.GET_RSSI:

        msg_contents = MessageToDict(msg_wrapper.rssi_msg)
    return msg_type, msg_contents    

File: utils/ProtobufTransformer.cpp
#include "ProtobufTransformer.hpp"

std::string encode_sys_msg(const std::string &whoami, const std::string &status)
{
    std::string serialized_data;
    Wrapper wrap_msg;
    wrap_msg.set_type(SYSTEM);
    System* sys_msg = wrap_msg.mutable_sys_msg();
    sys_msg->set_whoami(whoami);
    sys_msg->set_status(status);
    wrap_msg.SerializeToString(&serialized_data);
    return serialized_data;
}

std::string encode_info_msg(const std::string &info)
{
    std::string serialized_data;
    Wrapper wrap_msg;
    wrap_msg.set_type(INFORMATION);
    Information* info_msg = wrap_msg.mutable_info_msg();
    info_msg->set_info(info);
    wrap_msg.SerializeToString(&serialized_data);
    return serialized_data;
}

std::string encode_motors_msg(const float alpha, const float beta)
{
    std::string serialized_data;
    Wrapper wrap_msg;
    wrap_msg.set_type(MOTORS);
    Motors* motors_msg = wrap_msg.mutable_motors_msg();
    motors_msg->set_alpha(alpha);
    motors_msg->set_beta(beta);
    wrap_msg.SerializeToString(&serialized_data);
    return serialized_data;
}

std::string encode_tx_msg(const float freq, const float power, const std::string &wavetype, const bool onoff)
{
    std::string serialized_data;
    Wrapper wrap_msg;
    wrap_msg.set_type(TRANSMIT);
    Transmit* tx_msg = wrap_msg.mutable_tx_msg();
    tx_msg->set_frequency(freq);
    tx_msg->set_power(power);
    tx_msg->set_wavetype(wavetype);
    tx_msg->set_onoff(onoff);
    wrap_msg.SerializeToString(&serialized_data);
    return serialized_data;
}

std::string encode_freq_msg(const float freq)
{
    std::string serialized_data;
    Wrapper wrap_msg;
    wrap_msg.set_type(SET_FREQUENCY);
    SetFrequency* freq_msg = wrap_msg.mutable_freq_msg();
    freq_msg->set_frequency(freq);
    wrap_msg.SerializeToString(&serialized_data);
    return serialized_data;
}

std::string encode_sample_msg(const bool onoff)
{
    std::string serialized_data;
    Wrapper wrap_msg;
    wrap_msg.set_type(SAMPLING);
    Sampling* sample_msg = wrap_msg.mutable_sample_msg();
    sample_msg->set_onoff(onoff);
    wrap_msg.SerializeToString(&serialized_data);
    return serialized_data;
}

std::string encode_rssi_msg(const float rssi, const int samples, const float std)
{
    std::string serialized_data;
    Wrapper wrap_msg;
    wrap_msg.set_type(GET_RSSI);
    GetRssi* rssi_msg = wrap_msg.mutable_rssi_msg();
    rssi_msg->set_rssi(rssi);
    rssi_msg->set_samples(samples);
    rssi_msg->set_std(std);
    wrap_msg.SerializeToString(&serialized_data);
    return serialized_data;
}

int get_type(const std::string &serialized_data)
{
    Wrapper wrap_msg;
    wrap_msg.ParseFromString(serialized_data);
    return wrap_msg.type();
}

Sys_struct decode_sys_msg(const std::string &serialized_data)
{
    Sys_struct sys;
    Wrapper wrap_msg;
    wrap_msg.ParseFromString(serialized_data);
    const System& msg = wrap_msg.sys_msg();
    sys.whoami = msg.whoami();
    sys.status = msg.status();
    return sys;
}

Info_struct decode_info_msg(const std::string &serialized_data)
{
    Info_struct information;
    Wrapper wrap_msg;
    wrap_msg.ParseFromString(serialized_data);
    const Information& msg = wrap_msg.info_msg();
    information.info = msg.info();
    return information;
}

Motors_struct decode_motors_msg(const std::string &serialized_data)
{
    Motors_struct motors;
    Wrapper wrap_msg;
    wrap_msg.ParseFromString(serialized_data);
    const Motors& msg = wrap_msg.motors_msg();
    motors.alpha = msg.alpha();
    motors.beta = msg.beta();
    return motors;
}

Tx_struct decode_tx_msg(const std::string &serialized_data)
{
    Tx_struct tx;
    Wrapper wrap_msg;
    wrap_msg.ParseFromString(serialized_data);
    const Transmit& msg = wrap_msg.tx_msg();
    tx.freq = msg.frequency();
    tx.power = msg.power();
    tx.wavetype = msg.wavetype();
    tx.onoff = msg.onoff();
    return tx;
}

Freq_struct decode_freq_msg(const std::string &serialized_data)
{
    Freq_struct setfreq;
    Wrapper wrap_msg;
    wrap_msg.ParseFromString(serialized_data);
    const SetFrequency& msg = wrap_msg.freq_msg();
    setfreq.freq = msg.frequency();
    return setfreq;
}

Sample_struct decode_sample_msg(const std::string &serialized_data)
{
    Sample_struct sample;
    Wrapper wrap_msg;
    wrap_msg.ParseFromString(serialized_data);
    const Sampling& msg = wrap_msg.sample_msg();
    sample.onoff = msg.onoff();
    return sample;
}

Rssi_struct decode_rssi_msg(const std::string &serialized_data)
{
    Rssi_struct getrssi;
    Wrapper wrap_msg;
    wrap_msg.ParseFromString(serialized_data);
    const GetRssi& msg = wrap_msg.rssi_msg();
    getrssi.rssi = msg.rssi();
    getrssi.samples =  msg.samples();
    getrssi.std =  msg.std();
    return getrssi;
}
